### DISCUSSION TITLE
Getting different trace Id when calling @Async calls in Java, using open telemetry manual instrumentation

### DISCUSSION THREAD

**Discussion by keerthanacv** [open-telemetry/opentelemetry-java-instrumentation#12112] [Q&A] [answered]
I'm using java manual instrumentation of open telemetry but receiving different trace Id even after using **scope** of open telemetry while using async calls

My Open telemetry configuration class

@Configuration
public class OpentelemetryConfiguration {
    @Bean
    public static OpenTelemetry openTelemetry() {
        Resource resource = Resource.getDefault();
        SdkTracerProvider sdkTracerProvider = SdkTracerProvider.builder()
                .addSpanProcessor(SimpleSpanProcessor.create(LoggingSpanExporter.create()))
                .setResource(resource)
                .build();

        SdkMeterProvider sdkMeterProvider = SdkMeterProvider.builder()
                .registerMetricReader(PeriodicMetricReader.builder(LoggingMetricExporter.create()).build())
                .setResource(resource)
                .build();

        SdkLoggerProvider sdkLoggerProvider = SdkLoggerProvider.builder()
                .addLogRecordProcessor(BatchLogRecordProcessor.builder(SystemOutLogRecordExporter.create()).build())
                .setResource(resource)
                .build();

        OpenTelemetry openTelemetry = OpenTelemetrySdk.builder()
                .setTracerProvider(sdkTracerProvider)
                .setMeterProvider(sdkMeterProvider)
                .setLoggerProvider(sdkLoggerProvider)
                .setPropagators(ContextPropagators.create(TextMapPropagator.composite(W3CTraceContextPropagator.getInstance(), W3CBaggagePropagator.getInstance())))
                .buildAndRegisterGlobal();

        return openTelemetry;
    }

    @Bean
    public Tracer tracer(OpenTelemetry openTelemetry) {
        return openTelemetry.getTracer("CommCentral Application");
    }

}

package com.unp.service;

import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.Tracer;
import org.slf4j.MDC;
import org.springframework.stereotype.Service;
import org.apache.commons.lang3.ObjectUtils;

@Service
public class OpentelemetryService {
    public Span controllerSpan(String spanName, Tracer tracer, String requestID, String traceID) {
        Span span = tracer.spanBuilder(spanName).startSpan();
        MDC.put("correlationId", span.getSpanContext().getTraceId());
        MDC.put("spanId", span.getSpanContext().getSpanId());
        MDC.put("requestId", requestID);
        MDC.put("class", spanName);
        MDC.put("traceId", traceID);
        return span;
    }

    public Span serviceSpan(String spanName, Tracer tracer) {
        Span serviceSpan = tracer.spanBuilder(spanName).startSpan();
        MDC.put("correlationId", serviceSpan.getSpanContext().getTraceId());
        MDC.put("spanId", serviceSpan.getSpanContext().getSpanId());
        MDC.put("class", spanName);
        return serviceSpan;
    }

    public void endSpan(Span span)
    {
        if(ObjectUtils.isNotEmpty(span))
            span.end();
    }
}

I have an endpoint /getData in Controller
Controller :

@Autowired
	private Tracer tracer;
	
	
public String getDate(Request request)
{
Span span = opentelemetryService.controllerSpan("Controller", tracer,
                (String) headers.get("request-id"), (String) headers.get("trace-id"));
		try(Scope scope = serviceSpan.makeCurrent()) {
service.getDataFromService();

} 
}

Service :

public String getDataFromService(Request request)
{
return getAsyncData1()+getAsyncData2();
}

Component:
@Autowired
	private Tracer tracer;

 @Async("AsyncExecutor")
 String getAsyncData1()
{
 Span serviceSpan = opentelemetryService.serviceSpan("CompletableFutureHandler",tracer);
        try(Scope scope = serviceSpan.makeCurrent())
return "1";
}
}

@Async("AsyncExecutor")
 String getAsyncData2()
{

 Span serviceSpan = opentelemetryService.serviceSpan("CompletableFutureHandler",tracer);
        try(Scope scope = serviceSpan.makeCurrent()){
return "2";
}
}

how to fix it?

**Comment by laurit** [comment, marked as answer]
When using manual instrumentation there is no automatic context propagation between threads, you'll have to take care of it on your own. One way to achieve this would be to wrap the executor using https://github.com/open-telemetry/opentelemetry-java/blob/7522bfe141524d52649b5c344c643104b57af608/context/src/main/java/io/opentelemetry/context/Context.java#L119

**Comment by keerthanacv** [comment]
Adding it the executor worked. Thank you!! @laurit

**Comment by birbirsbirbirs** [comment]
can you share the sample code please. how to wrap default applicationTaskExecutor? Thank you very very much.

**Reply by keerthanacv** [reply]
@Override
    @Bean(name = "AsyncExecutor")
    public Executor getAsyncExecutor() {
        ThreadPoolTaskExecutor taskExecutor = new ThreadPoolTaskExecutor();
        taskExecutor.setCorePoolSize(poolSize);
        taskExecutor.setMaxPoolSize(maxPoolSize);
        taskExecutor.setQueueCapacity(threadPoolQueueCapacity);
        taskExecutor.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
        taskExecutor.setThreadNamePrefix("Executor-");
        taskExecutor.initialize();
        return taskWrapping(taskExecutor); // we need send this
    }
    
    // the below part helped
    public Executor taskWrapping(ThreadPoolTaskExecutor executor) {
        return command -> executor.execute(Context.current().wrap(command));
    }
